Detectors/FOCAL/simulation/src/Detector.cxx

```diff
@@ -11,30 +11,83 @@
 
 #include <TParticle.h>
 #include <TVirtualMC.h>
+#include <TGeoVolume.h>
+#include <TGeoManager.h>
+#include <TGeoBBox.h>
+#include <TGeoCompositeShape.h>
 
 #include <FairVolume.h>
 
+#include "DetectorsBase/Stack.h"
 #include "FOCALSimulation/Detector.h"
 
 using namespace o2::focal;
 
+Detector::Detector(bool active, std::string geofilename)
+  : o2::base::DetImpl<Detector>("FOC", active),
+    mHits(o2::utils::createSimVector<Hit>()),
+    mGeometry(nullptr),
+    mMedSensHCal(-1),
+    mGeoCompositions(0x0),
+    mSuperParentsIndices(),
+    mSuperParents(),
+    mCurrentSuperparent(nullptr),
+    mCurrentTrack(-1),
+    mCurrentPrimaryID(-1),
+    mCurrentParentID(-1),
+    mVolumeIDScintillator(-1)
+{
+  mGeometry = getGeometry(geofilename);
+}
+
 Detector::Detector(const Detector& rhs)
+  : o2::base::DetImpl<Detector>(rhs)
 {
+  mGeometry = rhs.mGeometry;
+  mMedSensHCal = rhs.mMedSensHCal;
+  mSensitiveHCAL = rhs.mSensitiveHCAL;
+  mVolumeIDScintillator = rhs.mVolumeIDScintillator;
 }
 
 Detector::~Detector()
 {
   o2::utils::freeSimVector(mHits);
 }
 
+Geometry* Detector::getGeometry(std::string name)
+{
+  if (!mGeometry) {
+    mGeometry = Geometry::getInstance(name);
+  }
+  if (!mGeometry) {
+    LOG(error) << "Failure accessing geometry";
+  }
+  return mGeometry;
+}
+
 void Detector::InitializeO2Detector()
 {
+
+  LOG(info) << "Intializing FOCAL detector";
+
+  // All FOCAL volumes must be declared as sensitive, otherwise
+  // the decay chains are broken by volumes not processed in ProceeHits
+  for (const auto& child : mSensitiveHCAL) {
+    LOG(debug1) << "Adding sensitive volume " << child;
+    auto svolID = registerSensitiveVolumeAndGetVolID(child);
+    if (child == "ScintFiber" || child == "HScint") {
+      LOG(debug1) << "Adding ScintFiber/HScint volume as sensitive volume with ID " << svolID;
+      mVolumeIDScintillator = svolID;
+    }
+  }
+
+  mMedSensHCal = getMediumID(ID_SC);
 }
 
 Bool_t Detector::ProcessHits(FairVolume* v)
 {
-  int track = fMC->GetStack()->GetCurrentTrackNumber(),
-      directparent = fMC->GetStack()->GetCurrentParentTrackNumber();
+  int track = gMC->GetStack()->GetCurrentTrackNumber(),
+      directparent = gMC->GetStack()->GetCurrentParentTrackNumber();
```

Comment:
I would avoid using the gMC variable and stick to the VMC app provided by FairModule.

Reply:
Using fMC now instead.